File: 027-二叉搜索树与双向链表/027-Convert.py
# -*- coding:utf-8 -*-
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None
class Solution:
    def Convert(self, pRootOfTree):
        # write code here
        if pRootOfTree is None:
            return None
        last=self.cvt(pRootOfTree,None)
        while last.left is not None:
            last=last.left
        return last
    def cvt(self,root,lastnode):
        if root.left is not None:
            lastnode=self.cvt(root.left,lastnode)
        if lastnode is not None:
            lastnode.right=root
            root.left=lastnode
        lastnode=root
        if root.right is not None:
            lastnode=self.cvt(root.right,lastnode)
        return lastnode

File: 027-二叉搜索树与双向链表/027-Convert.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:
    TreeNode* Convert(TreeNode* pRootOfTree)
    {
        if(pRootOfTree ==nullptr){
            return nullptr;
        }
        TreeNode* last=cvt(pRootOfTree,nullptr);
        while(last->left!=nullptr){
            last=last->left;
        }
        return last;
    }
    TreeNode* cvt(TreeNode* root,TreeNode* lastnode){
        if (root->left!=nullptr){
            lastnode=cvt(root->left,lastnode);
        }
        if(lastnode!=nullptr){
            lastnode->right=root;
            root->left=lastnode;
        }
        lastnode=root;
        if (root->right!=nullptr){
            lastnode=cvt(root->right,lastnode);
        }
        return lastnode;
    }
};
